### ISSUE TITLE
Consider publishing to hex.pm?

### ISSUE THREAD

**Issue by reisub** [axelson/publicsuffix-elixir#2]
Hi @axelson, thanks for keeping this package alive and maintained!

I'm working on a new [HTTP Cookie package](https://github.com/reisub/http_cookie) which uses this functionality and would like to add an optional dependency to it, and to do that it needs to be published.

For now I'm mentioning the dep in the readme so it's not a big deal, but I remember you mentioning you would consider publishing it.

Are you still maintaining this package and do you want to publish it?
I can help with maintenance as well.

Thanks in advance.

**Comment by axelson** [comment]
Hi @reisub 👋!

Publishing it is definitely still on the table. But if possible I'd like to see the original GitHub repository transitioned as well so that there's minimal disruption to the users of the project. I've started a discussion at https://github.com/orgs/beam-community/discussions/23 about potentially organizing maintenance of the library under the Beam Community umbrella.

**Comment by reisub** [comment]
I've tried to be patient with this, but a recent change in how optional dependencies are resolved with hex forced my hand; I've published a fork of this library as [public_sufx](https://hex.pm/packages/public_sufx) and plan to maintain it going forward.

@axelson thanks for helping try to get the original repo transitioned.

I'm happy to accept contributions to public_sufx.
Feel free to close this issue.

**Comment by axelson** [comment]
Thank you @reisub, I am glad to see the library moving forward. I've responded to the other thread about potentially keeping the existing package name on hex.pm.

By the way do you have a link to the optional dependency change? I'm not aware of it.

**Comment by reisub** [comment]
I don't have a link to the change so I'm not sure what exactly changed or when. @wojtekmach mentioned something about mix pruning code paths between compilations so avoid depending on compilation order.

So then I tried to use the code from earlier version of the lib (implementation was something like this, the idea being you can only use the functionality if you have the public_suffix lib):

```elixir
if Code.ensure_loaded?(PublicSuffix) do
  def public_suffix?(domain) do
    domain == PublicSuffix.public_suffix(domain)
  end
else
  def public_suffix?(_domain) do
    raise "Missing :public_suffix dependency"
  end
end
```

with the dependencies set up like this:

```
{:http_cookie, "~> 0.7.0"},
# needed for http_cookie public suffix check
{:public_suffix, github: "axelson/publicsuffix-elixir"},
```

and I saw what he meant, the public suffix functionality was broken (raising) **sometimes**.
My guess is that if the compilation order is PublicSuffix -> HttpCookie then it works, otherwise not; but I've not verified this.

@wojtekmach do you remember more about this?

**Comment by wojtekmach** [comment]
> `if Code.ensure_loaded?(PublicSuffix) do`

where does this code live? If in http_cookie package then it needs to have a dependency on phblic_suffix, can be an optional dependency.

**Comment by reisub** [comment]
> > `if Code.ensure_loaded?(PublicSuffix) do`
> 
> where does this code live? If in http_cookie package then it needs to have a dependency on phblic_suffix, can be an optional dependency.

Yeah, it used to live in `http_cookie` and I wanted to make `axelson/publicsuffix-elixir` an optional dep but then I couldn't build a package for publishing with a github dep - the mix task wouldn't let me.

I initially thought just adding a dependency without the lib declaring an optional dependency would work. I tested it at the time and it worked, but in hindsight it probably only sometimes worked.

So to resolve this mess, I opted for publishing a fork.

With HttpCookie 0.8.0 it's no longer an optional dependency because disabling it can lead to security issues and I don't want to encourage this. Of course, if it turns out there is a legit reason to disable the public suffix check, we can add an option for that in a future release.

**Comment by wojtekmach** [comment]
If it is a regular ie non optional dependency then that Code.ensure_loaded? should not be needed anymore. If you can reproduce any weird behavior I am happy to take a look!

**Comment by reisub** [comment]
> If it is a regular ie non optional dependency then that Code.ensure_loaded? should not be needed anymore. If you can reproduce any weird behavior I am happy to take a look!

Yeah, all good. It's not needed and it's not there anymore for 0.8.0.
AFAICT it works as expected now - thanks for the offer though! 🙏

**Comment by axelson** [comment]
Ah, that makes sense. And with the publishing of your library I'm going to close this issue and move over to your fork for all of my public suffix needs. Thank you for publishing it!

**Comment by axelson** [comment]
@reisub can you enable GitHub issues on https://github.com/reisub/public_sufx. By default issues are not enabled on forks.

Also I'd like to contend for setting `download_data_on_compile: false` as the default. In general (in my opinion which of course you are free to ignore!) libraries should avoid making network requests during compile time by default for a few reasons:
- making a network request means the build is not reproducible
  - and if the request fails then potentially the build will fail (which could cause CI to fail)
- it can cause the upstream servers to be "unfairly" hammered, for example this is what is happening with tzdata https://github.com/lau/tzdata/issues/149 (and that's one reason I generally use and recommend using `tz` over `tzdata`)
  - if this is the default then many users will not disable this downloading in their CI pipelines
- it is generally not expected by most consumers of the library

Instead I'd bundle the public suffix list with `public_sufx` directly or as a separate package that can be updated and released independently. That will give users more control over what version to use and they can auto update with a tool like dependabot.

Also if you haven't seen it before you might be interested in https://github.com/zachdaniel/git_ops which makes it much less effort to generate each release.

**Comment by reisub** [comment]
> [@reisub](https://github.com/reisub) can you enable GitHub issues on https://github.com/reisub/public_sufx. By default issues are not enabled on forks.

Thanks for the note, enabled.

> Also I'd like to contend for setting `download_data_on_compile: false` as the default. In general (in my opinion which of course you are free to ignore!) libraries should avoid making network requests during compile time by default for a few reasons:
> 
> * making a network request means the build is not reproducible
>   
>   * and if the request fails then potentially the build will fail (which could cause CI to fail)
> * it can cause the upstream servers to be "unfairly" hammered, for example this is what is happening with tzdata [please don't hammer on https://data.iana.org lau/tzdata#149](https://github.com/lau/tzdata/issues/149) (and that's one reason I generally use and recommend using `tz` over `tzdata`)
>   
>   * if this is the default then many users will not disable this downloading in their CI pipelines
> * it is generally not expected by most consumers of the library
> 
> Instead I'd bundle the public suffix list with `public_sufx` directly or as a separate package that can be updated and released independently. That will give users more control over what version to use and they can auto update with a tool like dependabot.

I was already on the fence about this.

I initially went for download by default because I expect there's lots of project that don't update dependencies very often so at face value it seems like a safer option.

That being said, you make some very good points.
The strongest one for me is the build reproducibility and the possibility of build failure.

I've opened an issue for this and will consider it:
https://github.com/reisub/public_sufx/issues/1

> Also if you haven't seen it before you might be interested in https://github.com/zachdaniel/git_ops which makes it much less effort to generate each release.

Thanks for taking the time to write this up, it's much appreciated!

**Comment by wojtekmach** [comment]
You may wanna check out https://github.com/elixir-mint/castore if you haven’t, especially their GitHub actions automation. Seems maybe related to what you are discussing.

**Comment by axelson** [comment]
Great, thank you!

**Comment by reisub** [comment]
Hey @axelson, [public_sufx](https://hex.pm/packages/public_sufx) has been published for a while now and I've set up some automation to make it easier to keep the shipped list up to date (thanks for the suggestion @wojtekmach!).

If you are happy with public_sufx and don't intend to maintain this repo anymore, would you consider adding a link to public_sufx in the README and archiving the repo?

**Comment by axelson** [comment]
@reisub yeah, no problem! I've archived it in https://github.com/axelson/publicsuffix-elixir/pull/3 Thank you for maintaining a fork! 💜
